### PR TITLE
Correctly return int values where required.

### PR DESCRIPTION
**Pull request by ghost**
### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/159
https://github.com/magento/partners-magento2b2b/pull/61

### Description (*)

The methods `getCustomerId` and `getCustomerGroupId` incorrectly return strings due to the usage of `getData`. This PR casts the values to ints, to match the return type hints.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by ihor-sviziev — DISMISSED** [review]
@paul-gene,
Your changes looks good, however there are some test failures, looks like your changes broke something.
Could you fix found issues?

**Comment by paul-gene** [comment]
@ihor-sviziev are the tests failing due to my actual code change?

**Comment by ihor-sviziev** [comment]
@paul-gene seems like yes, but let’s restart failing tests, maybe it was some issue with testing infra
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

**Comment by paul-gene** [comment]
@ihor-sviziev If my code change is the cause for failing tests, then the problem is bigger than I thought, as it means tests were written to be broken from the start. fml.

Can the tests be re-run without a commit to kick them into action?

**Comment by ihor-sviziev** [comment]
@paul-gene TBH it's not seems the change chat could break functionality, but it might in not obvious way. All functional and integration tests are running all the time for all PRs, and I didn't saw that these tests were failing somewhere else.

**Comment by paul-gene** [comment]
@ihor-sviziev I can see a few PRs around the same time as mine that are failing the same tests.

**Comment by paul-gene** [comment]
@ihor-sviziev there must be something wrong with the tests. Look at this one that failed

**Comment by paul-gene** [comment]
The "expected" result looks to be empty in the test results. I don't see how my change could have broken that?

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-6947](https://jira.corp.magento.com/browse/ENGCOM-6947) has been created to process this Pull Request
:eight_spoked_asterisk: @VladimirZaets, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by magento-engcom-team** [comment]
@paul-gene thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by paul-gene** [comment]
@VladimirZaets what's the next step to get this merged?

**Comment by paul-gene** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

@paul-gene this command available only for maintainers

**Comment by paul-gene** [comment]
Thanks @ihor-sviziev 

Are you able to explain this though? -https://github.com/magento/magento2/pull/26930#issuecomment-588966295

From what I can tell (allure is terrible to read), there is no possible way my code change could be breaking some of those tests.

**Comment by ihor-sviziev** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, here is your [new Magento instance](https://pr-26930.instances.magento-community.engineering).
Admin access: [https://pr-26930.instances.magento-community.engineering/admin_1f8a](https://pr-26930.instances.magento-community.engineering/admin_1f8a)
Login: `895cd15a` Password: `7103a389a7de`
Instance will be terminated in up to 3 hours.

**Review by ihor-sviziev — DISMISSED** [review]
@paul-gene I was able to reproduce on test instance that case with not filled up region for Belgium:
image

Failed test link:
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/26930/ccd81e7e427e0ddc9df6410896bce9c2/Functional/allure-report-ce/index.html#categories/a35b298dfea20498553ff2fcc5bf7ccb/34dc9351db11a9a3/

I think this is the reason why the tests are failing.

Could you debug how your solution and fix the found issue?

**Review by ihor-sviziev — DISMISSED** [review]
See comment above

**Comment by paul-gene** [comment]
@ihor-sviziev the region looks set to me in that screenshot?

Again, if you actually look at my code change, it is casting the customer ID to an integer, as it should have been from the beginning.

**Comment by ihor-sviziev** [comment]
@paul-gene I absolutely agree with you, it's not logical at all. But some issues really not logical. I was able to reproduce this regression on test instance, while it worked fine on `2.4-develop` branch, so I believe you will be able to reproduce it locally.

All steps to reproduce you could see in failed test result, you could open groups to see all steps that were executed:
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/26930/ccd81e7e427e0ddc9df6410896bce9c2/Functional/allure-report-ce/index.html#categories/a35b298dfea20498553ff2fcc5bf7ccb/34dc9351db11a9a3/

**Comment by paul-gene** [comment]
@ihor-sviziev  you stated "I was able to reproduce on test instance that case with not filled up region for Belgium" yet in the test, the region ID is set as "Hainaut" and that appears in the screenshot.

Screen Shot 2020-02-25 at 15 02 53

Screen Shot 2020-02-25 at 15 03 28

**Comment by ihor-sviziev** [comment]
@paul-gene,
Just scrolled down, there is below editing of shipping address by changing region to "Please select a region, state or province." and saving it:
image

**Comment by paul-gene** [comment]
@ihor-sviziev then that is an issue with the test, as of course it will not save if you set the region to "Please select a region, state or province." as it's a null value.

It's worth nothing the test failing was added to 2.4-develop 4 months ago. I hazard it was not tested properly itself.

**Comment by ihor-sviziev** [comment]
@paul-gene 
why do you think that test isn't correct?
Region field is not mandatory (there is no stars) in magento for Belgium country, so you have to be able to save it:
image

For most countries you'll have required region, or you have to write it yourself, but for some - it's not restricted to be mandatory field. From my perspective it's absolutely valid case. Don't you think so?

**Comment by paul-gene** [comment]
@ihor-sviziev can you explain how casting the customer ID to an int would break this test?

The address is initially saving, so the problem lies with the region being missing. If the address saves OK the first time, then my code change is **not** the cause of the failure

**Comment by ihor-sviziev** [comment]
@engcom-Alfa @engcom-Golf 
could you help with test failures?

**Comment by paul-gene** [comment]
Paging @engcom-Alfa @engcom-Golf

**Comment by paul-gene** [comment]
Paging @engcom-Alfa @engcom-Golf

**Comment by paul-gene** [comment]
Paging @engcom-Alfa @engcom-Golf

**Comment by paul-gene** [comment]
Paging @engcom-Alfa @engcom-Golf

**Comment by paul-gene** [comment]
Paging @engcom-Alfa @engcom-Golf

File: app/code/Magento/Customer/Model/Session.php
Comment (Nazar65) [inline app/code/Magento/Customer/Model/Session.php:344]: This changes affects failed tests, this method should return null or int in your case it always return int, because (int)null == 0;

File: app/code/Magento/Customer/Model/Session.php
Comment (Nazar65) [inline app/code/Magento/Customer/Model/Session.php:395]: same here

**Review by Nazar65 — CHANGES_REQUESTED** [review]

File: app/code/Magento/Customer/Model/Session.php
Comment (ihor-sviziev) [reply app/code/Magento/Customer/Model/Session.php:344]: @Nazar65 Here is `if ($this->storage->getData('customer_id')) {` one line above, it should return null if the value is null. I believe it's not an issue

Comment (Nazar65) [reply app/code/Magento/Customer/Model/Session.php:344]: ah, i see agree

**Comment by engcom-Golf** [comment]
Hi @paul-gene i'm found the reason of failing test, give me some time i will update you branch

**Comment by paul-gene** [comment]
> Hi @paul-gene i'm found the reason of failing test, give me some time i will update you branch

I fixed it ;)

**Comment by engcom-Golf** [comment]
@paul-gene nope :) with your commit test still failing

**Comment by paul-gene** [comment]
> @paul-gene nope :) with your commit test still failing
> Problem in strict compare `=== `statements, where getCustomerId returns `'int'` not an` int`

check again, https://github.com/magento/magento2/pull/26930/commits/5a2e2ce243a5d60ba30e522b3d772827b0e7c60b

Your commit fixed a totally different file

**Comment by engcom-Golf** [comment]
@paul-gene exactly

File: app/code/Magento/Customer/Model/Address.php
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Address.php]: As I understood - there is no such method in parent. It have to be replaced with getData()

File: app/code/Magento/Customer/Controller/Adminhtml/Address/Delete.php
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Controller/Adminhtml/Address/Delete.php]: it's better to convert all the types to single type.
```suggestion
        if ($addressId && (int)$this->addressRepository->getById($addressId)->getCustomerId() === (int)$customerId) {
```

File: app/code/Magento/Customer/Model/Address.php
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Address.php]: In case if you'll create new model and will egecute this method, like this:
```php
$address = $this->addressFactory->create();
$address->getCustomerId();
```

Result will be like this: `(int) $this->getParentId();`  => `(int)null`  => `0`. I believe this is NOT expected result, it should be null.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @paul-gene,

@engcom-Golf investigated the issues in tests. in many places we have strict type comparison, so as result of your change - these places became working broken because `(int)'1' === '1' //false`, but `(int)'1' == '1' //true`. You could find  all these places by looking on failing tests and debugging.

Using actually strict comparison is a best practice, so fixing type is really correct! 👍

Could you review all failing tests and adjust all places that became broken?

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Hi @paul-gene,
Will you be able to review my comments in https://github.com/magento/magento2/pull/26930#pullrequestreview-371864758 and update your PR?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by paul-gene** [comment]
@ihor-sviziev yes, I read your comment. Work has been done by another contributor since then. Is there still something I need to do?

**Comment by ihor-sviziev** [comment]
@paul-gene I see that we have some test failures in B2B version where you don't have access to. So nothing from your side is needed at this point. 

Just merged with latest changes in 2.4-develop and will see if we'll have these failures.

**Comment by engcom-Golf** [comment]
@engcom-Echo will continue fixing failed b2b tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6947](https://jira.corp.magento.com/browse/ENGCOM-6947) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE

**Comment by paul-gene** [comment]
@ihor-sviziev looks like we're green. Can we get a merge going on? :)

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6947](https://jira.corp.magento.com/browse/ENGCOM-6947) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@paul-gene 
Now we’re waiting for approving all related PRs in b2b and EE repos

**Review by Nazar65 — DISMISSED** [review]

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

Tested functionality related to Create, Edit, Delete Customer Addresses, Quotes, Wishlist, and everything works as previously (expected).

**Review by slavvka — CHANGES_REQUESTED** [review]
@paul-gene Unfortunately we have an error in our performance sanity check. For some reason opening the customer edit page fails:
```
1587742244099,240,Edit Customer (Admin Customer Management),200,OK,One Thread Scenarios Pool 5-1,text,false,Response was null,112515,0,1,1,http://localhost/builds/backend/customer/index/edit/id/1/,236,0,0
```

Could you please check what is the reason of that. You need to run jMeter scenario against your code ([sanity (3).zip](https://github.com/magento/magento2/files/4530727/sanity.3.zip)).

More information you can find here https://github.com/magento/magento2ce/tree/2.4-develop/setup/performance-toolkit. There's a mistake in there, you need to take 3rd version

**Comment by ihor-sviziev** [comment]
Hi @paul-gene,
I don't see any changes, so not sure we do have something to review. Maybe you missed to push your changes?

**Comment by paul-gene** [comment]
@ihor-sviziev honestly, I don't understand any of @slavvka 's comment.

Everything was approved and all tests passed, so how all of a sudden did something break?

**Comment by ihor-sviziev** [comment]
@paul-gene your changes looks good, but we found that during running performance test we found performance degradation in one specific case (that @slavvka described above). You could reproduce it by running attached jMeter profile from and compare changes from your PR with 2.4-develop branch. 

This is one of the last step before merging PR, that's why it happened. 

Unfortunately we should strong motivation for accepting PRs that introduce performance degradation, while this PR don't have such, so it seems like PR should be adjusted.

Hope you got my point. If you have any questions 

BTW I found that in comment above we have incorrect link, here is correct one https://github.com/magento/magento2/tree/2.4-develop/setup/performance-toolkit

**Comment by paul-gene** [comment]
@ihor-sviziev I'll be honest, a lot of work was done after my initial commits, and I've never heard of, or used, jMeter.

I'm not sure I believe that casting variables to integers would degrade performance enough for it to make any conceivable difference, if at all :/

**Comment by slavvka** [comment]
@engcom-Echo could you please help @paul-gene find out what's wrong with the PAT Sanity scenario?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by orlangur — CHANGES_REQUESTED** [review]
Please provide test scenarios in which returning int is required (i.e. broken functionality).

`int` type hint covers both integers and numeric strings in fact.

See https://github.com/magento/magento2/pull/14057#issuecomment-372644527 for more details.

**Comment by ihor-sviziev** [comment]
Hi @orlangur,
Issue appears when you're using strict types `declare(strict_types=1)` in your file. Then you expect that if value should be int - you can use it for your method that has declaration like this 
```php
<?php
declare(strict_types=1);

function myFunction(int $storeId): void
{
}

$storeId = '2'; // let's assume we executed $store->getId(); that has defined return type int
myFunction($storeId);
```

output of `php 1.php`:
```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to myFunction() must be of the type integer, string given, called in /home/ihor/1.php on line 7 and defined in /home/ihor/1.php:3
Stack trace:
#0 /home/ihor/1.php(7): myFunction('2')
#1 {main}
  thrown in /home/ihor/1.php on line 3
```


In this case you'll get type error. Why - because string !== int.

**Comment by orlangur** [comment]
@ihor-sviziev this code is incorrect, we cannot rely on types derived from files without strict types declaration.

@paul-gene as per discussion in maintainers chat I'm closing this PR. We must use strict types in all new code, however, we have too many numeric strings in legacy code and any such change may cause bugs.

**Pull request closed without merging**
